src/tonic/api/raft_api.rs

```diff
@@ -19,11 +28,58 @@ impl Raft for RaftService {
             .map_err(|err| {
                 Status::invalid_argument(format!("Failed to parse raft message: {err}"))
             })?;
-        self.0
+        self.message_sender
             .lock()
             .map_err(|_| Status::internal("Can't capture the Raft message sender lock"))?
             .send(consensus::Message::FromPeer(Box::new(message)))
             .map_err(|_| Status::internal("Can't send Raft message over channel"))?;
         Ok(Response::new(()))
     }
+
+    async fn who_is(
+        &self,
+        request: tonic::Request<PeerId>,
+    ) -> Result<tonic::Response<UriStr>, tonic::Status> {
+        let addresses = self
+            .toc
+            .peer_address_by_id()
+            .map_err(|err| Status::internal(format!("Can't get peer addresses: {err}")))?;
+        let uri = addresses
+            .get(&request.get_ref().id)
+            .ok_or_else(|| Status::internal("Peer not found"))?;
+        Ok(Response::new(UriStr {
+            uri: uri.to_string(),
+        }))
+    }
+
+    async fn add_peer_to_known(
+        &self,
+        request: tonic::Request<Peer>,
+    ) -> Result<tonic::Response<AllPeers>, tonic::Status> {
+        let peer = request.into_inner();
+        self.toc
+            .propose_consensus_op(ConsensusOperations::AddPeer(peer.id, peer.uri), None)
+            .await
+            .map_err(|err| Status::internal(format!("Failed to add peer: {err}")))?;
+        let addresses = self
+            .toc
+            .peer_address_by_id()
+            .map_err(|err| Status::internal(format!("Can't get peer addresses: {err}")))?;
+        Ok(Response::new(AllPeers {
+            all_peers: addresses
+                .into_iter()
+                .map(|(id, uri)| Peer {
+                    id,
+                    uri: uri.to_string(),
+                })
+                .collect(),
+        }))
+    }
+
+    async fn add_peer_as_participant(
```

Comment:
It would be nice to have a comment here - what is the difference with `add_peer_to_known`, and where to use which

Reply:
There is a comment in the protobuf service definition and trait declaration, I just did not include it here in the impl.

Reply:
I can add it here also if you think it is important.

Reply:
Yeah, this section is rather complicated, I would prefer to have a bit more comments